### DISCUSSION TITLE
0.36.3

### DISCUSSION THREAD

**Discussion by rmorshea** [idom-team/idom#680] [Announcements]
Misc bug fixes along with a minor improvement that allows components to return ``None``
to render nothing.

**Closed Issues**

- All child states wiped upon any child key change - #652 
- Allow NoneType returns within components - #538

**Merged Pull Requests**

- fix #652 - #672
- Fix 663 -  #667

<hr /><em>This discussion was created from the release <a href='https://github.com/idom-team/idom/releases/tag/0.36.3'>0.36.3</a>.</em>

**Comment by Archmonger** [comment]
Thanks, I'll test this out soon!

**Comment by Archmonger** [comment]
@rmorshea This version is critically broken. Layout rendering is now completely non-functional on Conreq/app_store with no visible errors.

**Reply by rmorshea** [reply]
Can you provide more detail?

**Reply by rmorshea** [reply]
Have you tried upgrading idom-client-react too?

**Reply by rmorshea** [reply]
Now that I think about it, I think this should have been a 0.37.0 release.

**Reply by Archmonger** [reply]
I had my venv which was the latest `django-idom`, and then I manually did `pip install idom --upgrade`. That action bumped the IDOM version from 0.36.0 to 0.36.3.

**Reply by Archmonger** [reply]
I just confirmed a completely fresh venv has the same behavior. At least for me, 0.36.3 is critically broken.

**Reply by rmorshea** [reply]
Did you also bump the JS version?

**Reply by rmorshea** [reply]
I put up an MR to bump the JS dependency in idom-django

**Comment by Archmonger** [comment]
The child state issue is resolved. I also did a quick test of using the `_` function as an [alternative to top level divs](https://github.com/idom-team/idom/issues/538), but this didn't work.

**Reply by rmorshea** [reply]
As in `html._` raises an error or it doesn't actually remove a level from the tree?

**Reply by Archmonger** [reply]
All VDOM contained within `html._` are not rendered. I'm guessing it has something to do with IDOM adding in it's own `html._` node within the layout render?

**Reply by rmorshea** [reply]
Hmm. This simple example seems to work as expected:

```python
from idom import component, html, run

@component
def Demo():
    return html._(
        html.h1("hello"),
        html.h1("world"),
    )

run(Demo)
```

I think something else might be going on. Does replacing `html._` with a `html.div` change the behavior?

**Reply by Archmonger** [reply]
Yes it works with `html.div`. I'm currently testing around to see if I can find situations within Conreq where `html._` works as intended.

**Reply by Archmonger** [reply]
**Does not work** in the following locations:

- [Base homepage component](https://github.com/Archmonger/Conreq/blob/36498d9a86e0c6fcaf8277cdbf0c6b9d3dc2619a/conreq/_core/home/components/__init__.py#L24)
   - The whole page simply doesn't get rendered if I utilize it here.
   - I use `idom.web.module_from_template` for my modals, and that does not play nicely with `html._`
   - If I remove any code that utilizes `idom.web.export` within my component, everything works again
- [Sidebar element](https://github.com/Archmonger/Conreq/blob/36498d9a86e0c6fcaf8277cdbf0c6b9d3dc2619a/conreq/_core/home/components/sidebar.py#L219)
   - Only doesn't work well here because `html._` does not support `key=...`. I can refactor my code around this, but is this limitation intentional?

**Works** in the following location:
- [App store viewport](https://github.com/Archmonger/Conreq/blob/36498d9a86e0c6fcaf8277cdbf0c6b9d3dc2619a/conreq/_core/app_store/components.py#L41)

**Reply by rmorshea** [reply]
Hadn't thought about keys. Can you create an issue about supporting keys in fragments?

**Reply by Archmonger** [reply]
I don't know if you caught my comment edit, but I determined that `html._` does not play nicely with things exported via `module_from_template`

**Reply by rmorshea** [reply]
And there's not client-side error of any kind?

Either way though, when you find time, if you can make a script that reproduces the issues above I'll try and debug it this coming week.

**Reply by Archmonger** [reply]
Nope, no client side error. I'll create a reproducible example now.

**Reply by Archmonger** [reply]
After some more testing I determined this was an issue on my side.

I should've inspected the DOM. Things are rendering. It failed previously due to some CSS that was made with the assumption of a top-level div. The BS modal created a div, which messed things up.

**Comment by Archmonger** [comment]
I've detected a new warning on console. I'm not sure how to reproduce. All I was doing was flipping through sidebar tabs.

```
Warning: Can't perform a React state update on an unmounted component. This is a no-op, but it indicates a memory leak in your application. To fix, cancel all subscriptions and asynchronous tasks in a useEffect cleanup function.
    at ImportedElement (http://127.0.0.1:7575/static/django_idom/client.js:31116:28)
    at Element (http://127.0.0.1:7575/static/django_idom/client.js:31050:20)
```

Let me know if I should generate an issue.

**Reply by Archmonger** [reply]
@rmorshea Let me know if this console log means anything to you.

**Reply by rmorshea** [reply]
Ok, I think I know what's going on here. This wasn't introduced by this release, but it's a bug. I can try to write it up.

**Reply by rmorshea** [reply]
https://github.com/idom-team/idom/issues/690
